src/pybind/mgr/volumes/fs/volume.py

```diff
@@ -287,6 +291,29 @@ def create_subvolume(self, **kwargs):
             ret = self.volume_exception_to_retval(ve)
         return ret
 
+    def _update_purge_stats(self, volname, fs_handle, group, subvolname):
+        '''
+        Count total number of files and subvolumes in trash directory, this
+        statistic is useful in reporting the progress made by purge threads.
+        '''
+        num_of_trash_files = get_subvol_stats(self.mgr, fs_handle, self.volspec, group,
+                                              subvolname)
+
+        if not self.purge_stats.get(volname, None):
+            self.purge_stats[volname] = {'total_files': num_of_trash_files,
+                                         'total_subvols': 1}
+            return
+
+        try:
+            self.purge_stats[volname]['total_files'] += num_of_trash_files
+        except KeyError:
+            self.purge_stats[volname]['total_files'] = num_of_trash_files
```

Comment:
Suggestion: Perhaps more cleaner 
```
self.purge_stats[volname]['total_files'] = self.purge_stats[volname].get('total_files', 0) + num_of_trash_files
```

But more over does this exception handling required ? Since if `self.purge_stats[volname]` is `None`, the `total_files` and `total_subvols` are getting initialized above! I think we can directly increment without exception handling here.

Reply:
> Perhaps more cleaner

I find it harder to read, the phrase `self.purge_stats[volname]['total_files']` is too long and repetitive and therefore error prone in what you suggested.


> But more over does this exception handling required ?

This is just defensive programming. In case the dictionary has volume added to it but rest of keys are missing, the command will crash. This instance of defensive programming prevents that.

Reply:
> Why on earth would the value of self.purge_stats[volname]

Usually what happens is someone in future someone introduces code where those keys are not added which leads to a unnecessary bug afterwards. So, it's better to have an extra check as a matter of defensive programming. This is degree of defensive programming is common.

> If you are putting in blocks of code under try..except

No, no. It's a matter of leaving no scope for future bugs.

Reply:
> > Why on earth would the value of self.purge_stats[volname]
> 
> Usually what happens is someone in future someone introduces code where those keys are not added which leads to a unnecessary bug afterwards. So, it's better to have an extra check as a matter of defensive programming. This is degree of defensive programming is common.

Code is reviewed and tested before merge. There is no point solving all of that prematurely. Defensive programming has its part in certain places but not justified when the dictionary key is inititlized a couple of lines before.

Reply:
> > Why on earth would the value of self.purge_stats[volname]
> 
> Usually what happens is someone in future someone introduces code where those keys are not added which leads to a unnecessary bug afterwards. So, it's better to have an extra check as a matter of defensive programming. This is degree of defensive programming is common.

I am not convinced. I think it's the job the someone to fix it correctly then especially for this use case.

Reply:
Cool, dropped.